In [1]:
import os
from pathlib import Path

# 항상 PRISM 루트로 고정
ROOT = Path('__file__').resolve().parent.parent if '__file__' in globals() else Path.cwd()
while ROOT.name != 'PRISM':
    ROOT = ROOT.parent
os.chdir(ROOT)
print(f'작업 디렉토리: {os.getcwd()}')

작업 디렉토리: c:\Users\rose0\OneDrive\Desktop\2026\대학교\1학기\캡스톤디자인(1)\PRISM


In [2]:
import pandas as pd

df = pd.read_csv('data/raw/fragella_raw.csv', low_memory=False)
print(f'총 행수: {len(df)}')
print(f'컬럼 수: {len(df.columns)}')
df.head(3)

총 행수: 101031
컬럼 수: 24


,_id,Name,Brand,Year,rating,Country,Image URL,Image URL Transparent,Gender,Price,...,Popularity,Price Value,General Notes,Main Accords,Main Accords Percentage,Season Ranking,Occasion Ranking,Notes,Image Fallbacks,Purchase URL
0,Seychelles-De-Parfum-Fanm,Seychelles De Parfum Fanm,Coco De Mer Collection,NaN,4.06,NaN,https://cdn.fragella.com/images/seychelles-de-...,https://cdn.fragella.com/images/seychelles-de-...,women,119.99,...,Low,okay,"['Coco De Mer', 'Jasmine', 'Gardenia', 'Saffro...","['patchouli', 'vanilla', 'white floral', 'gour...","{'patchouli': 'Dominant', 'vanilla': 'Dominant...","[{'name': 'winter', 'score': 3.281}, {'name': ...","[{'name': 'night out', 'score': 1.22}, {'name'...","{'Top': [{'name': 'saffron', 'imageUrl': 'http...",['https://cdn.fragrancenet.com/images/photos/6...,https://www.fragrancenet.com/perfume/coco-de-m...
1,Seychelles-De-Parfum-Coco-De-Mer-Oud,Seychelles De Parfum Coco De Mer Oud,Coco De Mer Collection,NaN,3.90,NaN,https://cdn.fragella.com/images/seychelles-de-...,https://cdn.fragella.com/images/seychelles-de-...,women,119.99,...,Low,okay,"['Coco De mMer', 'Oud Sandalwoood', 'Cahsmere'...","['patchouli', 'vanilla', 'white floral', 'wood...","{'patchouli': 'Dominant', 'vanilla': 'Dominant...","[{'name': 'winter', 'score': 3.281}, {'name': ...","[{'name': 'night out', 'score': 1.22}, {'name'...","{'Top': [{'name': 'lemon', 'imageUrl': 'https:...",['https://cdn.fragrancenet.com/images/photos/6...,https://www.fragrancenet.com/perfume/coco-de-m...
2,Seychelles-De-Parfum-Lodas,Seychelles De Parfum Lodas,Coco De Mer Collection,NaN,NaN,NaN,https://cdn.fragella.com/images/seychelles-de-...,https://cdn.fragella.com/images/seychelles-de-...,unisex,119.99,...,Low,okay,"['Coco De Mer', 'Yuzu', 'Lemongrass', 'Pink Pe...","['citrus', 'soft spicy', 'fresh spicy', 'herba...","{'citrus': 'Dominant', 'soft spicy': 'Dominant...","[{'name': 'spring', 'score': 1.271}, {'name': ...","[{'name': 'professional', 'score': 1.33}, {'na...","{'Top': [{'name': 'lemongrass', 'imageUrl': 'h...",['https://cdn.fragrancenet.com/images/photos/6...,https://www.fragrancenet.com/fragrances/coco-d...


In [3]:
# 결측률 확인
print('=== 컬럼별 결측률 ===')
for col in df.columns:
    null_n = df[col].isna().sum()
    null_pct = null_n / len(df) * 100
    print(f'{col}: {null_n}개 ({null_pct:.1f}%)')

=== 컬럼별 결측률 ===
_id: 0개 (0.0%)
Name: 0개 (0.0%)
Brand: 0개 (0.0%)
Year: 37401개 (37.0%)
rating: 13288개 (13.2%)
Country: 70348개 (69.6%)
Image URL: 0개 (0.0%)
Image URL Transparent: 0개 (0.0%)
Gender: 118개 (0.1%)
Price: 43886개 (43.4%)
OilType: 50993개 (50.5%)
Longevity: 0개 (0.0%)
Sillage: 0개 (0.0%)
Confidence: 2개 (0.0%)
Popularity: 0개 (0.0%)
Price Value: 2개 (0.0%)
General Notes: 0개 (0.0%)
Main Accords: 0개 (0.0%)
Main Accords Percentage: 0개 (0.0%)
Season Ranking: 0개 (0.0%)
Occasion Ranking: 0개 (0.0%)
Notes: 0개 (0.0%)
Image Fallbacks: 1개 (0.0%)
Purchase URL: 32344개 (32.0%)


In [4]:
# 컬럼 타입 확인
print(df.dtypes)

_id                         object
Name                        object
Brand                       object
Year                        object
rating                     float64
Country                     object
Image URL                   object
Image URL Transparent       object
Gender                      object
Price                      float64
OilType                     object
Longevity                   object
Sillage                     object
Confidence                  object
Popularity                  object
Price Value                 object
General Notes               object
Main Accords                object
Main Accords Percentage     object
Season Ranking              object
Occasion Ranking            object
Notes                       object
Image Fallbacks             object
Purchase URL                object
dtype: object


In [5]:
# 주요 컬럼 분포 확인
print('=== Year 분포 (상위 20개) ===')
print(df['Year'].value_counts().head(20))

print('\n=== Gender 분포 ===')
print(df['Gender'].value_counts())

print('\n=== Popularity 분포 ===')
print(df['Popularity'].value_counts())

print('\n=== Confidence 분포 ===')
print(df['Confidence'].value_counts())

=== Year 분포 (상위 20개) ===
Year
2015    4384
2019    4314
2018    4232
2016    4141
2020    4109
2017    3948
2022    3835
2021    3633
2014    3185
2023    3113
2013    2950
2012    2902
2010    1940
2011    1939
2024    1809
2009    1659
2008    1489
2006    1181
2007    1053
2005     721
Name: count, dtype: int64

=== Gender 분포 ===
Gender
women     41683
unisex    39883
men       18762
Men         584
Unisex        1
Name: count, dtype: int64

=== Popularity 분포 ===
Popularity
Low            46522
Medium         23539
Not popular    19143
High            8767
Very high       3060
Name: count, dtype: int64

=== Confidence 분포 ===
Confidence
low       73685
high      14768
medium    12576
Name: count, dtype: int64


In [6]:
# Year 연대별 분포
df['Year_num'] = pd.to_numeric(df['Year'], errors='coerce')

bins   = [1800, 1900, 1950, 1970, 1980, 1990, 2000, 2005, 2010, 2015, 2020, 2027]
labels = ['~1900', '1900s', '1950s', '1970s', '1980s', '1990s', '2000~04', '2005~09', '2010~14', '2015~19', '2020~']

era = pd.cut(df['Year_num'], bins=bins, labels=labels)
print('=== 연대별 향수 수 ===')
print(era.value_counts().sort_index())
print(f'\nMin: {df["Year_num"].min():.0f}, Max: {df["Year_num"].max():.0f}')
print(f'Unknown: {df["Year_num"].isna().sum()}개 ({df["Year_num"].isna().sum()/len(df)*100:.1f}%)')

df.drop(columns=['Year_num'], inplace=True)

=== 연대별 향수 수 ===
Year_num
~1900         97
1900s        522
1950s        364
1970s        539
1980s        978
1990s       2637
2000~04     2673
2005~09     7322
2010~14    15360
2015~19    20744
2020~      12390
Name: count, dtype: int64

Min: 1800, Max: 2024
Unknown: 37402개 (37.0%)


In [ ]:
# 중복 현황 확인
print(f'전체 행수: {len(df)}')
print(f'_id 기준 중복: {df.duplicated(subset=["_id"]).sum()}개')
print(f'Name+Brand 기준 중복: {df.duplicated(subset=["Name", "Brand"]).sum()}개')

print('\n=== 중복이 많은 브랜드 상위 10개 ===')
dup_by_brand = df[df.duplicated(subset=['_id'], keep=False)].groupby('Brand').size()
print(dup_by_brand.sort_values(ascending=False).head(10))

전체 행수: 101031
_id 기준 중복: 62303개
Name+Brand 기준 중복: 62356개

=== 중복이 많은 브랜드 상위 10개 ===
Brand
Anna Sui          683
Afnan Perfumes    537
Acqua Di Parma    531
Paul Smith        434
Natura            430
By Kilian         420
essence           404
Nike              361
Franck Olivier    357
Michel Germain    355
dtype: int64


In [8]:
# 중복 원인 분석 - Anna Sui 케이스
import json

with open('data/raw/collection_log.json', 'r', encoding='utf-8') as f:
    log = json.load(f)

anna = df[df['Brand'] == 'Anna Sui']
print(f'Anna Sui 전체 행수: {len(anna)}')
print(f'Anna Sui 고유 향수 수 (_id 기준): {anna["_id"].nunique()}')
print(f'collection_log 수집 기록: {log["brand_counts"].get("Anna Sui", 0)}개 (1회 호출)')

id_counts = anna['_id'].value_counts()
print(f'\n중복 횟수 — 최소: {id_counts.min()}회 / 최대: {id_counts.max()}회 / 평균: {id_counts.mean():.1f}회')

print('\n=== 중복 횟수별 향수 수 ===')
print('중복 횟수  향수 수')
for cnt, n in id_counts.value_counts().sort_index().items():
    print(f'{cnt:<10} {n}')

Anna Sui 전체 행수: 683
Anna Sui 고유 향수 수 (_id 기준): 50
collection_log 수집 기록: 50개 (1회 호출)

중복 횟수 — 최소: 2회 / 최대: 23회 / 평균: 13.7회

=== 중복 횟수별 향수 수 ===
중복 횟수  향수 수
2          1
4          1
5          2
7          4
10         8
11         5
13         5
14         2
15         4
18         9
19         1
20         3
21         2
22         1
23         2


In [9]:
# Anna Sui 그룹 분석
anna_indices = sorted(df[df['Brand'] == 'Anna Sui'].index.tolist())

chunks = []
chunk_start = anna_indices[0]
prev = anna_indices[0]

for idx in anna_indices[1:]:
    if idx - prev > 100:
        chunks.append((chunk_start, prev))
        chunk_start = idx
    prev = idx
chunks.append((chunk_start, prev))

print(f'Anna Sui 그룹 수: {len(chunks)}개 (collection_log 호출 기록: 1회)')
for i, (start, end) in enumerate(chunks):
    size = len([x for x in anna_indices if start <= x <= end])
    print(f'  그룹 {i+1}: 인덱스 {start}~{end}, {size}개')

Anna Sui 그룹 수: 14개 (collection_log 호출 기록: 1회)
  그룹 1: 인덱스 5531~5564, 34개
  그룹 2: 인덱스 6235~6633, 304개
  그룹 3: 인덱스 6742~6783, 42개
  그룹 4: 인덱스 6892~6983, 84개
  그룹 5: 인덱스 7129~7183, 8개
  그룹 6: 인덱스 7890~7911, 22개
  그룹 7: 인덱스 38028~38060, 24개
  그룹 8: 인덱스 42764~42810, 18개
  그룹 9: 인덱스 49410~49456, 18개
  그룹 10: 인덱스 72208~72209, 2개
  그룹 11: 인덱스 75001~75020, 9개
  그룹 12: 인덱스 83406~83550, 26개
  그룹 13: 인덱스 89540~89586, 46개
  그룹 14: 인덱스 93105~93154, 46개


**Anna Sui를 예시로 분석한 결과, 'collection_log 기준 1회 호출 / 50개 수집'임에도
실제 데이터에는 '14개 그룹 / 683행'이 존재한다.**

그룹 크기가 2~304개로 제각각이며, 직접 호출 시 매번 동일한 50개가 반환되는 것과 다르다.

이는 다른 브랜드 호출 시 Anna Sui 향수가 응답에 섞여 들어온 것으로 판단된다.

재수집해도 동일하게 발생하므로, 전처리 단계에서 `_id` 기준 중복 제거로 처리하기로 했다.